ehrql/query_engines/base_sql.py

```diff
@@ -77,19 +78,39 @@ def get_query(self, variable_definitions):
         `get_setup_and_cleanup_queries` on the query object.
         """
         variable_definitions = apply_transforms(variable_definitions)
+
+        # Generate a table containing the IDs all of patients matching the population
+        # definition
+        population_expression = self.get_predicate(variable_definitions["population"])
+        select_patient_id = self.select_patient_id_for_population(population_expression)
+        population_query = select_patient_id.where(population_expression)
+        population_query = apply_patient_joins(population_query)
+        population_table = self.reify_query(population_query)
+        # Store a reference to the population table so that we can use it while
+        # generating the variable expressions below
+        self.population_table = population_table
+
         variable_expressions = {
             name: self.get_expr(definition)
             for name, definition in variable_definitions.items()
             if name != "population"
         }
-        population_expression = self.get_predicate(variable_definitions["population"])
-        query = self.select_patient_id_for_population(population_expression)
+        query = sqlalchemy.select(population_table.c.patient_id)
         query = query.add_columns(
             *[expr.label(name) for name, expr in variable_expressions.items()]
         )
-        query = query.where(population_expression)
         query = apply_patient_joins(query)
 
+        # We use an instance variable to store the population table in order to avoid
+        # having to thread it through all our `get_sql`/`get_table` method calls. But
+        # this means that we can't safely re-use cached values across different calls to
+        # `get_query()` because the same query node may now generate different SQL
+        # depending on the population of the query to which it belongs. So we have to
+        # reset the caches and the population table reference.
+        self.population_table = None
+        self.get_sql.cache_clear()
+        self.get_table.cache_clear()
```

Comment:
None of the tests exercise the clearing of the caches, so if we add further cached single dispatch methods, there's a chance we'll miss this.  Can we try to hit this in a test somehow?  And can we add comments to `get_sql` and `get_table` about when the caches need to be cleared, as a clue to future us.

Or what about extracting most of the methods on `BaseSQLQueryEngine` into a `BaseSQLCompiler` class?  Then in `BaseSQLQueryEngine.get_query`, we instantiate a new compiler, and throw it away at the end of the method.